Lib/ast.py

```diff
@@ -1497,6 +1504,26 @@ def visit_withitem(self, node):
             self.write(" as ")
             self.traverse(node.optional_vars)
 
+    def visit_match_case(self, node):
+        self.fill("case ")
+        self.traverse(node.pattern)
+        if node.guard:
+            self.write(" if ")
+            self.traverse(node.guard)
+        with self.block():
+            self.traverse(node.body)
+
+    def visit_MatchAs(self, node):
+        with self.require_parens(_Precedence.TEST, node):
+            self.set_precedence(_Precedence.EXPR, node.pattern)
+            self.traverse(node.pattern)
+            self.write(f" as {node.name}")
+
+    def visit_MatchOr(self, node):
+        with self.require_parens(_Precedence.EXPR, node):
+            self.set_precedence(_Precedence.ATOM, *node.patterns)
+            self.interleave(lambda: self.write(" | "), self.traverse, node.patterns)
+
```

Comment:
I'm not 100% how the operator precedence machinery works here, but this feels correct to me and behaves as expected. I think `TEST`, `EXPR`, and `ATOM` are quite arbitrary, but the order is correct, so whatever. 🙂

Reply:
For the `MatchOr`, it could only require the parentheses on `_Precedence.BOR` (same with `EXPR`, but more explicit) (I would also go for changing the `EXPR` in the `MatchAs` with the `BOR`). For the patterns inside of binary-or, we could simply mark them with the next precedence that comes right after the `BOR`, which is `BXOR` (so that we could avoid adding parentheses around the complex numbers).

Reply:
How's something like this?

```suggestion
    def visit_MatchAs(self, node):
        with self.require_parens(_Precedence.TEST, node):
            self.set_precedence(_Precedence.BOR, node.pattern)
            self.traverse(node.pattern)
            self.write(f" as {node.name}")

    def visit_MatchOr(self, node):
        with self.require_parens(_Precedence.BOR, node):
            self.set_precedence(_Precedence.BOR.next(), *node.patterns)
            self.interleave(lambda: self.write(" | "), self.traverse, node.patterns)

```

I like `_Precedence.BOR.next()` better than `_Precedence.BXOR`

Reply:
...well that's a stupid diff.

Reply:
> I think I like _Precedence.BOR.next() better than _Precedence.BXOR.

That's clever! (diff LGTM)